libsql-client/src/connection.rs

```diff
@@ -0,0 +1,61 @@
+//! `Connection` is the main structure to interact with the database.
+
+use async_trait::async_trait;
+
+use anyhow::Result;
+
+use super::{QueryResult, Statement};
+
+/// Trait describing capabilities of a database connection:
+/// - executing statements, batches, transactions
+#[async_trait(?Send)]
+pub trait Connection {
+    /// Executes a single SQL statement
+    ///
+    /// # Arguments
+    /// * `stmt` - the SQL statement
+    async fn execute(&self, stmt: impl Into<Statement>) -> Result<QueryResult> {
+        let mut results = self.batch(std::iter::once(stmt)).await?;
+        Ok(results.remove(0))
+    }
+
+    /// Executes a batch of SQL statements.
+    /// Each statement is going to run in its own transaction,
+    /// unless they're wrapped in BEGIN and END
+    ///
+    /// # Arguments
+    /// * `stmts` - SQL statements
+    /// ```
+    async fn batch(
+        &self,
+        stmts: impl IntoIterator<Item = impl Into<Statement>>,
+    ) -> Result<Vec<QueryResult>>;
+
+    /// Executes an SQL transaction.
+    /// Does not support nested transactions - do not use BEGIN or END
+    /// inside a transaction.
+    ///
+    /// # Arguments
+    /// * `stmts` - SQL statements
+    /// ```
+    async fn transaction(
+        &self,
+        stmts: impl IntoIterator<Item = impl Into<Statement>>,
+    ) -> Result<Vec<QueryResult>> {
+        // TODO: Vec is not a good fit for popping the first element,
+        // let's return a templated collection instead and let the user
+        // decide where to store the result.
```

Comment:
Are we popping the first element? It seems that the first element is skipped and only the last element is popped, no?

Reply:
Is this comment still valid?

Reply:
nah, we're no longer using vectors, so not valid, I'll remove it